pkg/logs/sender/batch_strategy.go

```diff
@@ -6,143 +6,93 @@
-
-	batchStrategyExpVars.Set(fmt.Sprintf("%s_%d", pipelineName, pipelineID), expVars)
-
+func newBatchStrategyWithClock(inputChan chan *message.Message, outputChan chan *message.Payload, serializer Serializer, batchWait time.Duration, maxBatchSize int, maxContentSize int, pipelineName string, clock clock.Clock, contentEncoding ContentEncoding) Strategy {
 	return &batchStrategy{
+		inputChan:        inputChan,
+		outputChan:       outputChan,
 		buffer:           NewMessageBuffer(maxBatchSize, maxContentSize),
 		serializer:       serializer,
 		batchWait:        batchWait,
-		climit:           make(chan struct{}, maxConcurrent),
+		contentEncoding:  contentEncoding,
 		syncFlushTrigger: make(chan struct{}),
-		syncFlushDone:    make(chan struct{}),
 		pipelineName:     pipelineName,
-		expVars:          expVars,
 		clock:            clock,
 	}
-
-}
-
-func (s *batchStrategy) Flush(ctx context.Context) {
-	select {
-	case <-ctx.Done():
-		return
-	default:
-		s.syncFlushTrigger <- struct{}{}
-		<-s.syncFlushDone
-	}
 }
 
-func (s *batchStrategy) syncFlush(inputChan chan *message.Message, outputChan chan *message.Message, send func([]byte) error) {
-	defer func() {
-		s.flushBuffer(outputChan, send)
-		s.pendingSends.Wait()
-	}()
-	for {
-		select {
-		case m, isOpen := <-inputChan:
-			if !isOpen {
-				return
-			}
-			s.processMessage(m, outputChan, send)
-		default:
-			return
-		}
-	}
+// Stop flushes the buffer and stops the strategy
+func (s *batchStrategy) Stop() {
+	s.syncFlushTrigger <- struct{}{}
+	close(s.inputChan)
 }
 
 // Send accumulates messages to a buffer and sends them when the buffer is full or outdated.
-func (s *batchStrategy) Send(inputChan chan *message.Message, outputChan chan *message.Message, send func([]byte) error) {
-	flushTicker := s.clock.Ticker(s.batchWait)
-	defer func() {
-		s.flushBuffer(outputChan, send)
-		flushTicker.Stop()
-		s.pendingSends.Wait()
-	}()
-	var startIdle = time.Now()
-	for {
-		select {
-		case m, isOpen := <-inputChan:
-
-			if !isOpen {
-				// inputChan has been closed, no more payloads are expected
-				return
-			}
-			s.expVars.AddFloat(expVarIdleMsMapKey, float64(time.Since(startIdle)/time.Millisecond))
-			var startInUse = time.Now()
+func (s *batchStrategy) Start() {
 
-			s.processMessage(m, outputChan, send)
-
-			s.expVars.AddFloat(expVarInUseMapKey, float64(time.Since(startInUse)/time.Millisecond))
-			startIdle = time.Now()
-
-		case <-flushTicker.C:
-			// the first message that was added to the buffer has been here for too long, send the payload now
-			s.flushBuffer(outputChan, send)
-		case <-s.syncFlushTrigger:
-			s.syncFlush(inputChan, outputChan, send)
-			s.syncFlushDone <- struct{}{}
+	go func() {
+		flushTicker := s.clock.Ticker(s.batchWait)
+		defer func() {
+			s.flushBuffer(s.outputChan)
+			flushTicker.Stop()
+		}()
+		for {
+			select {
+			case m, isOpen := <-s.inputChan:
+
+				if !isOpen {
+					// inputChan has been closed, no more payloads are expected
```

Comment:
small question: Is this condition useful for debugging? If so, maybe a log message here would be nice.

Reply:
probably not. This only happens during graceful shutdown ([and hasn't changed in this PR](https://github.com/DataDog/datadog-agent/blob/main/pkg/logs/sender/batch_strategy.go#L117) - just moved :) )